WDL/runtime/backend/singularity.py

```diff
@@ -49,11 +67,16 @@ def cli_exe(self) -> List[str]:
     def _pull_invocation(self, logger: logging.Logger, cleanup: ExitStack) -> Tuple[str, List[str]]:
         image, invocation = super()._pull_invocation(logger, cleanup)
         docker_uri = "docker://" + image
-
-        # The docker image layers are cached in SINGULARITY_CACHEDIR, so we don't need to keep the
-        # *.sif
-        pulldir = cleanup.enter_context(tempfile.TemporaryDirectory(prefix="miniwdl_sif_"))
-        return (docker_uri, self.cli_exe + ["pull", "--dir", pulldir, docker_uri])
+        if self.image_cache_dir is not None:
+            image_name = docker_uri.replace("/", "_").replace(":", "_")
+            image_path = os.path.join(self.image_cache_dir, image_name)
+            if not os.path.exists(image_path):
+                return (image_path, self.cli_exe + ["pull", image_path, docker_uri])
+            # If path already exists test the image
+            return (image_path, self.cli_exe + ["exec", image_path, "true"])
+        # Singularity will cache the image automatically in
+        # SINGULARITY_CACHE_DIR it is not needed to save the image elsewhere.
+        return (docker_uri, self.cli_exe + ["exec", docker_uri, "true"])
```

Comment:
That is correct, it will just run the SIF file rather than use its own cache. I can rework the code so the image is always pulled, but in the default case the image is added to cleanup.